### ISSUE TITLE
Eval bug: GGML_ASSERT failure at ggml.c:6370 during llama_opt_epoch with Saiga Nemo 12B

### ISSUE THREAD

**Issue by jakexcosme** [COG-GTM/llama.cpp#178]
**Note: This issue was copied from [https://github.com/ggml-org/llama.cpp/issues/15279](https://github.com/ggml-org/llama.cpp/issues/15279)**

**Original Author:** @Recouper
**Original Issue Number:** #15279
**Created:** 2025-08-13T01:59:41Z

---

### Name and Version

.\llama-cli.exe --version
register_backend: registered backend CPU (1 devices)
register_device: registered device CPU (AMD Ryzen 5 5500                               )
load_backend: failed to find ggml_backend_init in C:\Users\Grant\llama.cpp\build\bin\Debug\ggml-cpu.dll
version: 6123 (79c1160b)
built with MSVC 19.44.35213.0 for x64

### Operating systems

Windows

### GGML backends

CPU, CUDA

### Hardware

AMD Ryzen 5 5500 + NVIDIA GeForce RTX 5060 Ti 16GB

### Models

saiga_nemo_12b.Q5_K_M

### Problem description & steps to reproduce

**Description**: Fine-tuning Saiga Nemo 12B (GGUF V3, Q5_K) with a 1290-token dataset crashes with `GGML_ASSERT(!node->view_src || node->op == GGML_OP_CPY || ...)` at `ggml.c:6370` during `llama_opt_epoch`.

**Reproduction**:
- Build: 6123 (79c1160b), MSVC 19.44.35213.0, CPU-only (-DGGML_CUDA=OFF -DLLAMA_CURL=OFF).
- Command: `llama-finetune.exe --model C:\Users\Grant\saiga_nemo_12b.Q5_K_M.gguf --train-data C:\Users\Grant\Documents\dataset_2.txt --output C:\Users\Grant\saiga_nemo_12b_finetuned.gguf --lora-out C:\Users\Grant\lora_adapter.bin --epochs 2 --lr 0.0002 --batch-size 2 --ctx-size 256 --no-warmup`
- Dataset: 20 Q&A pairs (~1290 tokens), `[INST] ... [/INST] ... [SEP]` format (attached or described below).

**System**: Windows, 12 threads, NVIDIA GeForce RTX 5060 Ti (16GB VRAM, unused in CPU mode).

**Attempts**: Tested `--ctx-size` (512, 256), `--batch-size` (1, 2), `--no-warmup`, CUDA and CPU modes—all failed at the same point.

**Observation**: `ndata = 8`, `idata_split = 7`, crash after `Starting epoch 1` and `Before llama_opt_epoch call`.

**Suspected Issue**: Possible bug in `ggml_opt_epoch` graph construction for Saiga Nemo 12B’s architecture (40 layers, 5120 embedding).

**Dataset Sample**:
[INST] What is the capital of France? [/INST] The capital of France is Paris. [SEP]
[INST] Explain gravity briefly. [/INST] Gravity is the force that attracts objects toward each other, like Earth pulling things down. [SEP]

**Request**: Advice on workaround or confirmation of bug.

### Relevant log output

```shell
main: force disabling memory mapping because it would result in-read-only pointers to the weights
main: force changing k cache type to f32 due to a lack of f16 support for OUT_PROD
main: force changing v cache type to f32 due to a lack of f16 support for OUT_PROD
build: 6123 (79c1160b) with MSVC 19.44.35213.0 for x64 (debug)
llama_model_loader: loaded meta data with 32 key-value pairs and 363 tensors from C:\Users\Grant\saiga_nemo_12b.Q5_K_M.gguf (version GGUF V3 (latest))
llama_model_loader: Dumping metadata keys/values. Note: KV overrides do not apply in this output.
llama_model_loader: - kv   0:                       general.architecture str              = llama
llama_model_loader: - kv   1:                               general.type str              = model
llama_model_loader: - kv   2:                               general.name str              = Saiga_Nemo_12B_Sft_M10_D16
llama_model_loader: - kv   3:                       general.organization str              = Models
llama_model_loader: - kv   4:                         general.size_label str              = 12B
llama_model_loader: - kv   5:                          llama.block_count u32              = 40
llama_model_loader: - kv   6:                       llama.context_length u32              = 1024000
llama_model_loader: - kv   7:                     llama.embedding_length u32              = 5120
llama_model_loader: - kv   8:                  llama.feed_forward_length u32              = 14336
llama_model_loader: - kv   9:                 llama.attention.head_count u32              = 32
llama_model_loader: - kv  10:              llama.attention.head_count_kv u32              = 8
llama_model_loader: - kv  11:                       llama.rope.freq_base f32              = 1000000.000000
llama_model_loader: - kv  12:     llama.attention.layer_norm_rms_epsilon f32              = 0.000010
llama_model_loader: - kv  13:                 llama.attention.key_length u32              = 128
llama_model_loader: - kv  14:               llama.attention.value_length u32              = 128
llama_model_loader: - kv  15:                          general.file_type u32              = 17
llama_model_loader: - kv  16:                           llama.vocab_size u32              = 131072
llama_model_loader: - kv  17:                 llama.rope.dimension_count u32              = 128
llama_model_loader: - kv  18:                       tokenizer.ggml.model str              = gpt2
llama_model_loader: - kv  19:                         tokenizer.ggml.pre str              = tekken
llama_model_loader: - kv  20:                      tokenizer.ggml.tokens arr[str,131072]  = ["<unk>", "<s>", "</s>", "[INST]", "[...
llama_model_loader: - kv  21:                  tokenizer.ggml.token_type arr[i32,131072]  = [3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, ...
llama_model_loader: - kv  22:                      tokenizer.ggml.merges arr[str,269443]  = ["─á ─á", "─á t", "e r", "i n", "─á ─...
llama_model_loader: - kv  23:                tokenizer.ggml.bos_token_id u32              = 1
llama_model_loader: - kv  24:                tokenizer.ggml.eos_token_id u32              = 2
llama_model_loader: - kv  25:            tokenizer.ggml.unknown_token_id u32              = 0
llama_model_loader: - kv  26:            tokenizer.ggml.padding_token_id u32              = 10
llama_model_loader: - kv  27:               tokenizer.ggml.add_bos_token bool             = true
llama_model_loader: - kv  28:               tokenizer.ggml.add_eos_token bool             = false
llama_model_loader: - kv  29:                    tokenizer.chat_template str              = {% if messages[0]['role'] == 'system'...
llama_model_loader: - kv  30:            tokenizer.ggml.add_space_prefix bool             = false
llama_model_loader: - kv  31:               general.quantization_version u32              = 2
llama_model_loader: - type  f32:   81 tensors
llama_model_loader: - type q5_K:  241 tensors
llama_model_loader: - type q6_K:   41 tensors
print_info: file format = GGUF V3 (latest)
print_info: file type   = Q5_K - Medium
print_info: file size   = 8.12 GiB (5.70 BPW)
load: special_eos_id is not in special_eog_ids - the tokenizer config may be incorrect
load: printing all EOG tokens:
load:   - 2 ('</s>')
load: special tokens cache size = 1000
load: token to piece cache size = 0.8498 MB
print_info: arch             = llama
print_info: vocab_only       = 0
print_info: n_ctx_train      = 1024000
print_info: n_embd           = 5120
print_info: n_layer          = 40
print_info: n_head           = 32
print_info: n_head_kv        = 8
print_info: n_rot            = 128
print_info: n_swa            = 0
print_info: is_swa_any       = 0
print_info: n_embd_head_k    = 128
print_info: n_embd_head_v    = 128
print_info: n_gqa            = 4
print_info: n_embd_k_gqa     = 1024
print_info: n_embd_v_gqa     = 1024
print_info: f_norm_eps       = 0.0e+00
print_info: f_norm_rms_eps   = 1.0e-05
print_info: f_clamp_kqv      = 0.0e+00
print_info: f_max_alibi_bias = 0.0e+00
print_info: f_logit_scale    = 0.0e+00
print_info: f_attn_scale     = 0.0e+00
print_info: n_ff             = 14336
print_info: n_expert         = 0
print_info: n_expert_used    = 0
print_info: causal attn      = 1
print_info: pooling type     = 0
print_info: rope type        = 0
print_info: rope scaling     = linear
print_info: freq_base_train  = 1000000.0
print_info: freq_scale_train = 1
print_info: n_ctx_orig_yarn  = 1024000
print_info: rope_finetuned   = unknown
print_info: model type       = 13B
print_info: model params     = 12.25 B
print_info: general.name     = Saiga_Nemo_12B_Sft_M10_D16
print_info: vocab type       = BPE
print_info: n_vocab          = 131072
print_info: n_merges         = 269443
print_info: BOS token        = 1 '<s>'
print_info: EOS token        = 2 '</s>'
print_info: UNK token        = 0 '<unk>'
print_info: PAD token        = 10 '<pad>'
print_info: LF token         = 1010 '─è'
print_info: EOG token        = 2 '</s>'
print_info: max token length = 150
load_tensors: loading model tensors, this can take a while... (mmap = false)
load_tensors:          CPU model buffer size =  8315.80 MiB
...........................................................................................
llama_context: constructing llama_context
llama_context: n_batch is less than GGML_KQ_MASK_PAD - increasing to 64
llama_context: n_seq_max     = 1
llama_context: n_ctx         = 256
llama_context: n_ctx_per_seq = 256
llama_context: n_batch       = 64
llama_context: n_ubatch      = 64
llama_context: causal_attn   = 1
llama_context: flash_attn    = 0
llama_context: kv_unified    = false
llama_context: freq_base     = 1000000.0
llama_context: freq_scale    = 1
llama_context: n_ctx_per_seq (256) < n_ctx_train (1024000) -- the full capacity of the model will not be utilized
llama_context:        CPU  output buffer size =     0.50 MiB
llama_kv_cache_unified:        CPU KV buffer size =    80.00 MiB
llama_kv_cache_unified: size =   80.00 MiB (   256 cells,  40 layers,  1/1 seqs), K (f32):   40.00 MiB, V (f32):   40.00 MiB
llama_context:        CPU compute buffer size =    38.56 MiB
llama_context: graph nodes  = 1406
llama_context: graph splits = 1
common_init_from_params: added </s> logit bias = -inf
common_init_from_params: setting dry_penalty_last_n to ctx_size = 256

system_info: n_threads = -1 / 12 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | LLAMAFILE = 1 | OPENMP = 1 | REPACK = 1 |
Loaded 1290 tokens from training data
Dataset initialized with ndata = 8
Starting epoch 1
Before llama_opt_epoch call, idata_split = 7
output_reserve: reallocating output buffer from size 0.50 MiB to 32.00 MiB
C:\Users\Grant\llama.cpp\ggml\src\ggml.c:6370: GGML_ASSERT(!node->view_src || node->op == GGML_OP_CPY || node->op == GGML_OP_VIEW || node->op == GGML_OP_RESHAPE || node->op == GGML_OP_PERMUTE || node->op == GGML_OP_TRANSPOSE) failed
```
